File: src/main/java/com/test/processor/subscribe/fanout/ReceiverA.java
package com.test.processor.subscribe.fanout;

import com.rabbitmq.client.*;
import com.test.util.ConnectionUtil;

import java.io.IOException;

/**
 * @author 费世程
 * @date 2020/4/7 17:33
 */
public class ReceiverA {

  private final static String QUEUE_NAME = "test_queue_subscribe_1";
  private final static String EXCHANGE_NAME = "test_exchange_fanout";

  public static void main(String[] args) {

    try {
      Connection connection = ConnectionUtil.Companion.getConnection();
      //获取通道
      Channel channel = connection.createChannel();
      //声明队列
      channel.queueDeclare(QUEUE_NAME, false, false, false, null);
      /*
        绑定队列到交换机
        String queue, String exchange, String routingKey
       */
      channel.queueBind(QUEUE_NAME, EXCHANGE_NAME, "");
      DefaultConsumer consumer = new DefaultConsumer(channel) {
        @Override
        public void handleDelivery(String consumerTag,
                                   Envelope envelope,
                                   AMQP.BasicProperties properties,
                                   byte[] body) throws IOException {
          String msg = new String(body, "UTF-8");
          System.out.println("receiver A :" + msg);
        }
      };
      channel.basicConsume(QUEUE_NAME, false, consumer);
    } catch (Exception e) {
      System.err.println("receiverA 接收消息出现异常：" + e);
    }

  }

}

File: src/main/java/com/test/processor/subscribe/direct/Receiver_A.kt
package com.test.processor.subscribe.direct

import com.rabbitmq.client.QueueingConsumer
import com.test.processor.subscribe.fanout.Subscribe_receiver_a
import com.test.util.ConnectionUtil

/**
 * @author 费世程
 * @date 2019/10/24 14:23
 */
class Receiver_A {

  companion object{
    private val QUEUE_NAME="test_queue_direct_1"
    private val EXCHANGE_NAME="test_exchange_direct"

    fun receive(){
      val connection=ConnectionUtil.getConnection()
      val channel=connection.createChannel()

      channel.queueDeclare(QUEUE_NAME,false,false,false,null)

      //绑定队列到交换机,同时指定需要订阅的routing key。可以指定多个
      //参数：String queue, String exchange, String routingKey
      channel.queueBind(QUEUE_NAME,EXCHANGE_NAME,"update")
      channel.queueBind(QUEUE_NAME,EXCHANGE_NAME,"delete")

      channel.basicQos(1)

      val consumer=QueueingConsumer(channel)
      //监听队列，手动返回完成
      channel.basicConsume(QUEUE_NAME,false,consumer)

      //获取消息
      while(true){
        val delivery=consumer.nextDelivery()
        val message=String(delivery.body)
        println("[receiver-a] : $message")
        Thread.sleep(10)
        channel.basicAck(delivery.envelope.deliveryTag,false)
      }
    }
  }

}